### ISSUE TITLE
Constant fold if statement with profiling information

### ISSUE THREAD

**Issue by jjsjann123** [csarofeen/pytorch#596]
## 🚀 Feature

`prim::If` statement creates blocks that prevents fusion of operation between nodes across the block. In most cases, the actual path taken at runtime rarely changes. So we would want to guard the conditional statement and inline the body to allow fusion of operations across blocks.

## Motivation

Two instances we are looking at:

a. `dropout` layer expanded by autodiff: https://github.com/csarofeen/pytorch/blob/27d5151be34af818dd9ddc110c7c44297ea45f5b/torch/csrc/jit/runtime/symbolic_script.cpp#L1051-L1079

b. `linear` layer in torch.nn: https://github.com/csarofeen/pytorch/blob/27d5151be34af818dd9ddc110c7c44297ea45f5b/torch/nn/functional.py#L1663-L1670

## Additional context

Things that we've tried/thought about:

The obvious thing that we tried is to utilize/modify `BailOut` node:

1. Prior to profiling run:
    i. Insert profile_ivalue node on all inputs to `prim::If` statement to record boolean value that determines to code path taken;
2. After profiling run, during optimization we apply a similar Guard/BailOut (This optimization should not run as an custom pass, but as part of pre-autodiff-pass-pipeline, because we can't expand fusion scope from within the fusion scope, during which custom pass are executed):
    i. Convert profile_ivalue node to modified BailOut node, where we compare input boolean to the recorded boolean value during profiling runs;
    ii. Replace input to `prim::If` with constant boolean created from profiled value; (Wiring `prim::BailOut` is tricky, we need a control-flow-edge from the node to inlined body, or we could inject a specialization rule to exclude it from DCE.)
    iii. Inline the body of `prim::If`;
3. At runtime:
    i. Modified BailOut node would check that the runtime conditional variable matches the profiled value, otherwise, it would bailout.

There's a limitation from this approach: BailOut node inserted for the check, which would still block fusion across itself.
In the example shown above, the conditional variables are meta information extracted from intermediate tensor (`use_cuda` and `dim()`) which can't be aggregated freely because of the dependency.

**Comment by csarofeen** [comment]
Closing as will not fix. We're moving towards program acquisition that won't have conditionals in the way of fusions. i.e. lazy tensors/lazy modes.
